File: 2024/08/solve.hs
#!/usr/bin/env runhaskell
{-# LANGUAGE TransformListComp #-}

import Data.List (nub, tails)
import GHC.Exts (groupWith)

type Point = (Int, Int)

parseInput :: String -> [[Point]]
parseInput input =
  [ p
  | (y, line) <- zip [0 ..] $ lines input,
    (x, c) <- zip [0 ..] line,
    c /= '.',
    let p = (x, y),
    then group by
      c
    using
      groupWith
  ]

squaredDistance :: Point -> Point -> Int
squaredDistance (x1, y1) (x2, y2) = (x2 - x1) * (x2 - x1) + (y2 - y1) * (y2 - y1)

-- | Determine colinearity of three points based on the area of the triangle formed
colinear :: Point -> Point -> Point -> Bool
colinear (x1, y1) (x2, y2) (x3, y3) = area == 0
  where
    area = x1 * (y2 - y3) + x2 * (y3 - y1) + x3 * (y1 - y2)

countAntinodes :: Bool -> Int -> [[Point]] -> Int
countAntinodes part2 dims bands = length $ nub antinodes
  where
    -- Generate all 2-combinations of a list
    pairs band = [(p1, p2) | (p1 : rest) <- tails band, p2 <- rest]
    pointPairs = concatMap pairs bands
    gridPoints = [(x, y) | x <- [0 .. dims - 1], y <- [0 .. dims - 1]]
    -- Check if point p3 satisfies the 1:2 distance requirement
    -- 4 is 2 squared btw
    distanceCriterion p1 p2 p3 =
      (squaredDistance p1 p3 == 4 * squaredDistance p2 p3)
        || (squaredDistance p2 p3 == 4 * squaredDistance p1 p3)
    antinodes =
      [ p3
      | p3 <- gridPoints,
        (p1, p2) <- pointPairs,
        colinear p1 p2 p3,
        part2 || distanceCriterion p1 p2 p3
      ]

main :: IO ()
main = do
  input <- readFile "input"
  let bands = parseInput input
      inputDims = length $ lines input
      partOne = countAntinodes False inputDims bands
      partTwo = countAntinodes True inputDims bands
  print partOne
  print partTwo

File: 2024/08/solve.py
#!/usr/bin/env python3
from collections import defaultdict
from itertools import combinations


def distance(point1: tuple[int, int], point2: tuple[int, int]):
    x1, y1 = point1
    x2, y2 = point2
    return ((x2 - x1) ** 2 + (y2 - y1) ** 2) ** 0.5


def colinear(p1, p2, p3):
    y1, x1 = p1
    y2, x2 = p2
    y3, x3 = p3
    area = x1 * (y2 - y3) + x2 * (y3 - y1) + x3 * (y1 - y2)
    return abs(area) < 1e-10


bands = defaultdict(list)
data = []
for y, line in enumerate(open("input")):
    data.append([])
    for x, c in enumerate(line.rstrip()):
        if c != ".":
            bands[c].append((y, x))
            data[-1].append(c)
        else:
            data[-1].append(c)


def count_antinodes(bands, data, distance_condition=True):
    antinodes = set()
    for coords in bands.values():
        pairs = list(combinations(coords, 2))
        for y in range(len(data)):
            for x in range(len(data[0])):
                point = (y, x)
                for point1, point2 in pairs:
                    if colinear(point1, point2, point):
                        if (
                            not distance_condition
                            or abs(
                                distance(point1, point) - 2 * distance(point2, (y, x))
                            )
                            < 0.005
                            or (
                                abs(
                                    distance(point2, point)
                                    - 2 * distance(point1, (y, x))
                                )
                                < 0.005
                            )
                        ):
                            antinodes.add((x, y))
    return len(antinodes)


print(part1 := count_antinodes(bands, data))
print(part2 := count_antinodes(bands, data, distance_condition=False))
